File: drouter-plugin/src/main/groovy/com/didi/drouter/utils/JarUtils.java
package com.didi.drouter.utils;

import java.io.BufferedReader;
import java.io.File;
import java.io.InputStreamReader;
import java.net.URL;
import java.util.Collection;
import java.util.Enumeration;
import java.util.HashMap;
import java.util.Map;
import java.util.jar.Attributes;
import java.util.jar.Manifest;
import java.util.zip.ZipEntry;
import java.util.zip.ZipFile;

/**
 * Created by gaowei on 2018/11/27.
 */
public class JarUtils {

    private static Map<String, String> metaInfo = new HashMap<>();

    private static void grabApiMetaInfo(Collection<File> files) {
        for (File file : files) {
            if (!file.getName().endsWith(".jar")) {
                continue;
            }
            try {
                ZipFile zipFile = new ZipFile(file);
                Enumeration<? extends ZipEntry> entries = zipFile.entries();
                while (entries.hasMoreElements()) {
                    ZipEntry entry = entries.nextElement();
                    if (!entry.isDirectory() && entry.getName().equals("META-INF/drouter")) {
                        BufferedReader reader =
                                new BufferedReader(new InputStreamReader(zipFile.getInputStream(entry)));
                        String line;
                        while ((line = reader.readLine()) != null) {
                            int index = line.indexOf(":");
                            if (index != -1) {
                                metaInfo.put(line.substring(0, index).trim(), line.substring(index + 1).trim());
                            }
                        }
                        reader.close();
                        break;
                    }
                }
            } catch (Exception e) {
                e.printStackTrace();
            }
        }
    }

    private static void grabPluginMetaInfo() {
        try {
            String classPath = JarUtils.class.getResource(JarUtils.class.getSimpleName() + ".class").toString();
            String libPath = classPath.substring(0, classPath.lastIndexOf("!"));
            String filePath = libPath + "!/META-INF/MANIFEST.MF";
            Manifest manifest = new Manifest(new URL(filePath).openStream());
            Attributes attributes = manifest.getMainAttributes();
            metaInfo.put("plugin-version", attributes.getValue("plugin-version"));
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    public static void printVersion(Collection<File> files) {
        grabApiMetaInfo(files);
        grabPluginMetaInfo();
        Logger.v("current plugin-version: " + metaInfo.get("plugin-version") +
                " | api-version: " + metaInfo.get("api-version"));
    }

    public static void check(Exception e) {
        Logger.w("[Wrong message ...]");
        Logger.e(">>> " + e.getMessage() + " <<<");
        String min = metaInfo.get("plugin-min-support");
        Logger.w("Please first make sure plugin-version " + metaInfo.get("plugin-version") +
                "(current) >= " + min + "(min support)");
        Logger.w("Any question you can also put in \"qq group 1017780981\"");
    }
}

File: plugins/drouter-plugin/src/main/kotlin/com/didi/drouter/utils/JarUtils.kt
package com.didi.drouter.utils

import com.android.build.gradle.AppExtension
import com.didi.drouter.utils.SystemUtil.project
import java.io.BufferedReader
import java.io.File
import java.io.InputStream
import java.io.InputStreamReader
import java.net.URL
import java.util.jar.Manifest
import java.util.zip.ZipFile

/**
 * Created by gaowei on 2018/11/27.
 */
object JarUtils {

    private val metaInfo = HashMap<String, String>()

    private fun grabApiMetaInfo(files: Collection<File>) {
        for (file in files) {
            if (!file.exists() || !file.name.endsWith(".jar")) {
                continue
            }
            val zipFile = ZipFile(file)
            try {
                val entries = zipFile.entries()
                while (entries.hasMoreElements()) {
                    val entry = entries.nextElement()
                    if (!entry.isDirectory && entry.name.equals("META-INF/drouter")) {
                        val reader = BufferedReader(InputStreamReader(zipFile.getInputStream(entry)))
                        var line = reader.readLine()
                        while (line != null) {
                            val index = line.indexOf(":")
                            if (index != -1) {
                                metaInfo[line.substring(0, index).trim()] = line.substring(index + 1).trim()
                            }
                            line = reader.readLine()
                        }
                        reader.close()
                        break
                    }
                }
            } catch (e: Exception) {
                e.printStackTrace()
            } finally {
                zipFile.close()
            }
        }
    }

    private fun grabPluginMetaInfo() {
        var inputStream: InputStream? = null
        try {
            val classPath = JarUtils::class.java.getResource(JarUtils::class.java.simpleName + ".class")!!.toString()
            val libPath = classPath.substring(0, classPath.lastIndexOf("!"))
            val filePath = "$libPath!/META-INF/MANIFEST.MF"
            inputStream = URL(filePath).openStream()
            val manifest = Manifest(inputStream)
            val attributes = manifest.mainAttributes
            metaInfo["plugin-version"] = attributes.getValue("Implementation-Version") ?: "null"
        } catch (e: Exception) {
            e.printStackTrace()
        } finally {
            inputStream?.close()
        }
    }

    @JvmStatic
    fun printVersion(files: Collection<File>) {
        grabApiMetaInfo(files)
        grabPluginMetaInfo()
        val pluginVersion = metaInfo["plugin-version"]
        val apiVersion = metaInfo["api-version"]
        Logger.v("current plugin-version: $pluginVersion | api-version: $apiVersion")

        val appIds = HashSet<String>()
        project.extensions.findByType(AppExtension::class.java)?.applicationVariants?.all { variant ->
            appIds.add(variant.applicationId)
        }
//        Thread.start {
//            sendRequest(Arrays.toString(appIds.toArray()), pluginVersion, apiVersion)
//        }
    }

    fun check(e: Exception) {
        Logger.w("[Wrong message ...]")
        Logger.e(">>> " + e.message + " <<<")
        val min = metaInfo["plugin-min-support"]
        Logger.w(
            "Please first make sure plugin-version " + metaInfo["plugin-version"] +
                    "(current) >= " + min + "(min support)"
        )
        Logger.w("Any question you can call \"GaoWei\" from weixin \"gwball\"")
    }

//    private static void sendRequest(String appId, String pluginVersion, String apiVersion) {
//        try {
//            def link = "https://czp.xiaojukeji.com/api/content/v1/compile?"
//            def args = "appId=" + URLEncoder.encode(appId, "UTF-8") +
//                    "&pluginVersion=" + URLEncoder.encode(pluginVersion == null ? "" : pluginVersion, "UTF-8") +
//                    "&apiVersion=" + URLEncoder.encode(apiVersion, "UTF-8")
//            def url = new URL(link + args)
//            def connection = (HttpURLConnection) url.openConnection()
//            connection.setRequestMethod("GET")
//            connection.setConnectTimeout(15000)
//            connection.setReadTimeout(60000)
//            def response = connection.inputStream.text
//        } catch(Exception ignore) {
//        }
//        //Logger.d("compile statistic response = " + response)
//    }
}
